File: Leetcode/Python/1780. Check if Number is a Sum of Powers of Three/solution.py
# https://leetcode.com/problems/check-if-number-is-a-sum-of-powers-of-three/description


class Solution:
    def checkPowersOfThree(self, n: int) -> bool:
        power_set = set()

        power = 0
        while 3 ** power <= n:
            power_set.add(3 ** power)
            power += 1

        power_values = list(power_set)
        power_values.sort()

        for idx in range(len(power_values) - 1, -1, -1):
            if power_values[idx] <= n:
                n -= power_values[idx]
                power_set.remove(power_values[idx])

        return n == 0

File: Leetcode/Typescript/1780. Check if Number is a Sum of Powers of Three/solution.ts
// https://leetcode.com/problems/check-if-number-is-a-sum-of-powers-of-three/description

function checkPowersOfThree(n: number): boolean {
  const powerSet = new Set<number>();

  let power = 0;

  while (3 ** power <= n) {
    powerSet.add(3 ** power);
    power += 1;
  }

  const powerValue = Array.from(powerSet.values());
  powerValue.sort((a, b) => a - b);

  for (let idx = powerValue.length - 1; idx >= 0; idx--) {
    if (powerValue[idx] <= n) {
      n -= powerValue[idx];
      powerSet.delete(n);
    }
  }

  return n == 0;
}
